src/test/crimson/test_messenger.cc

```diff
@@ -572,6 +578,1650 @@ seastar::future<> test_preemptive_shutdown(bool v2) {
+        }).then([&test] {
+          return test.peer_connect_me();
+        }).then([&suite] {
+          return suite.wait_results(1);
+        }).then([] (ConnResults& results) {
+          results[0].assert_state_at(conn_state_t::closed);
+          results[0].assert_connect(0, 0, 0, 0);
+          results[0].assert_accept(1, 1, 0, 1);
+          results[0].assert_reset(1, 0);
+        });
+      });
+    });
+  });
+}
+
+seastar::future<>
+test_v2_lossless_connect_fault(FailoverTest& test) {
+  return seastar::do_with(std::vector<Breakpoint>{
+      {Tag::CLIENT_IDENT, true},
+      {Tag::SERVER_IDENT, false},
+  }, [&test] (auto& failure_cases) {
+    return seastar::do_for_each(failure_cases, [&test] (auto bp) {
+      ostringstream sout;
+      sout << "test_v2_lossless_connect_fault -- " << bp;
+      TestInterceptor interceptor;
+      interceptor.make_fault(bp);
+      return test.run_suite(sout.str(),
+                            interceptor,
+                            policy_t::lossless_client,
+                            policy_t::stateful_server,
+                            [&test] (FailoverSuite& suite) {
+        return seastar::futurize_apply([&test] {
+          return test.send_bidirectional();
+        }).then([&suite] {
+          return suite.connect_peer();
+        }).then([&suite] {
+          return suite.wait_results(1);
+        }).then([] (ConnResults& results) {
+          results[0].assert_state_at(conn_state_t::established);
+          results[0].assert_connect(2, 2, 0, 1);
+          results[0].assert_accept(0, 0, 0, 0);
+          results[0].assert_reset(0, 0);
+        });
+      });
+    });
+  });
+}
+
+seastar::future<>
+test_v2_lossless_connected_fault(FailoverTest& test) {
+  return seastar::do_with(std::vector<Breakpoint>{
+      {Tag::MESSAGE, true},
+      {Tag::MESSAGE, false},
+  }, [&test] (auto& failure_cases) {
+    return seastar::do_for_each(failure_cases, [&test] (auto bp) {
+      ostringstream sout;
+      sout << "test_v2_lossless_connected_fault -- " << bp;
+      TestInterceptor interceptor;
+      interceptor.make_fault(bp);
+      return test.run_suite(sout.str(),
+                            interceptor,
+                            policy_t::lossless_client,
+                            policy_t::stateful_server,
+                            [&test] (FailoverSuite& suite) {
+        return seastar::futurize_apply([&test] {
+          return test.send_bidirectional();
+        }).then([&suite] {
+          return suite.connect_peer();
+        }).then([&suite] {
+          return suite.wait_results(1);
+        }).then([] (ConnResults& results) {
+          results[0].assert_state_at(conn_state_t::established);
+          results[0].assert_connect(2, 1, 1, 2);
+          results[0].assert_accept(0, 0, 0, 0);
+          results[0].assert_reset(0, 0);
+        });
+      });
+    });
+  });
+}
+
+seastar::future<>
+test_v2_lossless_reconnect_fault(FailoverTest& test) {
```

Comment:
I repeated this test for ~100 times, but can't reproduce this issue from my environment..
My passing logs look like:
```
[test_v2_lossless_reconnect_fault -- MESSAGE_READ, SESSION_RECONNECT_OK_READ]
INFO  2019-09-04 10:18:01,431 [shard 0] ms - [osd.2(Test) -] listening on 127.0.0.1:9010
INFO  2019-09-04 10:18:01,432 [shard 0] ms - [osd.4(TestPeer) -] listening on 127.0.0.1:9012
INFO  2019-09-04 10:18:01,433 [shard 0] ms - [0] [osd.2(Test) v2:127.0.0.1:9010/2 >> unknown.? -] new connection registered
INFO  2019-09-04 10:18:01,433 [shard 0] ms - [osd.2(Test) v2:127.0.0.1:9010/2 >> osd.? v2:127.0.0.1:9012/4] ProtocolV2::start_connect(): peer_addr=v2:127.0.0.1:9012/4, peer_type=osd, cc=8274171783502195587 policy(lossy=0, server=0, standby=0, resetcheck=1)
INFO  2019-09-04 10:18:01,433 [shard 0] ms - [0] [osd.2(Test) v2:127.0.0.1:9010/2 >> osd.? v2:127.0.0.1:9012/4] intercepted SOCKET_CONNECTING
INFO  2019-09-04 10:18:01,433 [shard 0] ms - Waiting for connection [0] [osd.2(Test) v2:127.0.0.1:9010/2 >> osd.? v2:127.0.0.1:9012/4] connected/closed
INFO  2019-09-04 10:18:01,433 [shard 0] ms - [osd.4(TestPeer) v2:127.0.0.1:9012/4 >> unknown.? -@63504] ProtocolV2::start_accept(): target_addr=127.0.0.1:63504/0
INFO  2019-09-04 10:18:01,433 [shard 0] ms - [0] [osd.2(Test) v2:127.0.0.1:9010/2 >> osd.? v2:127.0.0.1:9012/4] intercepted BANNER_WRITE
INFO  2019-09-04 10:18:01,433 [shard 0] ms - [0] [osd.2(Test) v2:127.0.0.1:9010/2 >> osd.? v2:127.0.0.1:9012/4] intercepted BANNER_READ
INFO  2019-09-04 10:18:01,433 [shard 0] ms - [0] [osd.2(Test) v2:127.0.0.1:9010/2 >> osd.? v2:127.0.0.1:9012/4] intercepted BANNER_PAYLOAD_READ
INFO  2019-09-04 10:18:01,433 [shard 0] ms - [0] [osd.2(Test) v2:127.0.0.1:9010/2 >> osd.? v2:127.0.0.1:9012/4] intercepted HELLO_WRITE
INFO  2019-09-04 10:18:01,433 [shard 0] ms - [osd.4(TestPeer) v2:127.0.0.1:9012/4 >> osd.? -@63504] UPDATE: peer_type=osd, policy(lossy=0 server=1 standby=1 resetcheck=1)
INFO  2019-09-04 10:18:01,433 [shard 0] ms - [0] [osd.2(Test) v2:127.0.0.1:9010/2 >> osd.? v2:127.0.0.1:9012/4] intercepted HELLO_READ
INFO  2019-09-04 10:18:01,433 [shard 0] ms - [0] [osd.2(Test) v2:127.0.0.1:9010/2@63504 >> osd.? v2:127.0.0.1:9012/4] intercepted AUTH_REQUEST_WRITE
INFO  2019-09-04 10:18:01,433 [shard 0] ms - [0] [osd.2(Test) v2:127.0.0.1:9010/2@63504 >> osd.? v2:127.0.0.1:9012/4] intercepted AUTH_DONE_READ
INFO  2019-09-04 10:18:01,433 [shard 0] ms - [0] [osd.2(Test) v2:127.0.0.1:9010/2@63504 >> osd.? v2:127.0.0.1:9012/4] intercepted AUTH_SIGNATURE_WRITE
INFO  2019-09-04 10:18:01,433 [shard 0] ms - [0] [osd.2(Test) v2:127.0.0.1:9010/2@63504 >> osd.? v2:127.0.0.1:9012/4] intercepted AUTH_SIGNATURE_READ
INFO  2019-09-04 10:18:01,433 [shard 0] ms - [0] [osd.2(Test) v2:127.0.0.1:9010/2@63504 >> osd.? v2:127.0.0.1:9012/4] intercepted CLIENT_IDENT_WRITE
INFO  2019-09-04 10:18:01,433 [shard 0] ms - [osd.4(TestPeer) v2:127.0.0.1:9012/4 >> osd.2 v2:127.0.0.1:9010/2@63504] accepted: gs=0, pgs=1, cs=0, client_cookie=8274171783502195587, server_cookie=1387656187526242379, in_seq=0, out_seq=0
INFO  2019-09-04 10:18:01,433 [shard 0] ms - [0] [osd.2(Test) v2:127.0.0.1:9010/2@63504 >> osd.? v2:127.0.0.1:9012/4] intercepted SERVER_IDENT_READ
INFO  2019-09-04 10:18:01,433 [shard 0] ms - [0] [osd.2(Test) v2:127.0.0.1:9010/2@63504 >> osd.4 v2:127.0.0.1:9012/4] got connected
INFO  2019-09-04 10:18:01,433 [shard 0] ms - [osd.2(Test) v2:127.0.0.1:9010/2@63504 >> osd.4 v2:127.0.0.1:9012/4] connected: gs=1, pgs=1, cs=0, client_cookie=8274171783502195587, server_cookie=1387656187526242379, in_seq=0, out_seq=0
INFO  2019-09-04 10:18:01,433 [shard 0] ms - [0] [osd.2(Test) v2:127.0.0.1:9010/2@63504 >> osd.4 v2:127.0.0.1:9012/4] ready
INFO  2019-09-04 10:18:01,433 [shard 0] ms - [0] [osd.2(Test) v2:127.0.0.1:9010/2@63504 >> osd.4 v2:127.0.0.1:9012/4] intercepted MESSAGE_READ
INFO  2019-09-04 10:18:01,433 [shard 0] ms - [0] [osd.2(Test) v2:127.0.0.1:9010/2 >> osd.4 v2:127.0.0.1:9012/4] intercepted SOCKET_CONNECTING
INFO  2019-09-04 10:18:01,433 [shard 0] ms - Waiting for connection [0] [osd.2(Test) v2:127.0.0.1:9010/2 >> osd.4 v2:127.0.0.1:9012/4] connected/closed
INFO  2019-09-04 10:18:01,434 [shard 0] ms - [0] [osd.2(Test) v2:127.0.0.1:9010/2 >> osd.4 v2:127.0.0.1:9012/4] intercepted BANNER_WRITE
INFO  2019-09-04 10:18:01,434 [shard 0] ms - [osd.4(TestPeer) v2:127.0.0.1:9012/4 >> unknown.? -@58824] ProtocolV2::start_accept(): target_addr=127.0.0.1:58824/0
INFO  2019-09-04 10:18:01,434 [shard 0] ms - [0] [osd.2(Test) v2:127.0.0.1:9010/2 >> osd.4 v2:127.0.0.1:9012/4] intercepted BANNER_READ
INFO  2019-09-04 10:18:01,434 [shard 0] ms - [0] [osd.2(Test) v2:127.0.0.1:9010/2 >> osd.4 v2:127.0.0.1:9012/4] intercepted BANNER_PAYLOAD_READ
INFO  2019-09-04 10:18:01,434 [shard 0] ms - [0] [osd.2(Test) v2:127.0.0.1:9010/2 >> osd.4 v2:127.0.0.1:9012/4] intercepted HELLO_WRITE
INFO  2019-09-04 10:18:01,434 [shard 0] ms - [0] [osd.2(Test) v2:127.0.0.1:9010/2 >> osd.4 v2:127.0.0.1:9012/4] intercepted HELLO_READ
INFO  2019-09-04 10:18:01,434 [shard 0] ms - [0] [osd.2(Test) v2:127.0.0.1:9010/2@58824 >> osd.4 v2:127.0.0.1:9012/4] intercepted AUTH_REQUEST_WRITE
INFO  2019-09-04 10:18:01,434 [shard 0] ms - [osd.4(TestPeer) v2:127.0.0.1:9012/4 >> osd.? -@58824] UPDATE: peer_type=osd, policy(lossy=0 server=1 standby=1 resetcheck=1)
INFO  2019-09-04 10:18:01,434 [shard 0] ms - [0] [osd.2(Test) v2:127.0.0.1:9010/2@58824 >> osd.4 v2:127.0.0.1:9012/4] intercepted AUTH_DONE_READ
INFO  2019-09-04 10:18:01,434 [shard 0] ms - [0] [osd.2(Test) v2:127.0.0.1:9010/2@58824 >> osd.4 v2:127.0.0.1:9012/4] intercepted AUTH_SIGNATURE_WRITE
INFO  2019-09-04 10:18:01,434 [shard 0] ms - [0] [osd.2(Test) v2:127.0.0.1:9010/2@58824 >> osd.4 v2:127.0.0.1:9012/4] intercepted AUTH_SIGNATURE_READ
INFO  2019-09-04 10:18:01,434 [shard 0] ms - [0] [osd.2(Test) v2:127.0.0.1:9010/2@58824 >> osd.4 v2:127.0.0.1:9012/4] intercepted SESSION_RECONNECT_WRITE
WARN  2019-09-04 10:18:01,434 [shard 0] ms - [osd.4(TestPeer) v2:127.0.0.1:9012/4 >> osd.? v2:127.0.0.1:9010/2@58824] server_reconnect: stale exsiting connect_seq exist_cs(0) < peer_cs(1), reusing existing [osd.4(TestPeer) v2:127.0.0.1:9012/4 >> osd.2 v2:127.0.0.1:9010/2@63504]
INFO  2019-09-04 10:18:01,434 [shard 0] ms - [osd.4(TestPeer) v2:127.0.0.1:9012/4 >> osd.2 v2:127.0.0.1:9010/2@63504] reconnected(replaced): gs=0, pgs=2, cs=1, client_cookie=8274171783502195587, server_cookie=1387656187526242379, in_seq=0, out_seq=0
WARN  2019-09-04 10:18:01,434 [shard 0] ms - [osd.4(TestPeer) v2:127.0.0.1:9012/4 >> osd.? v2:127.0.0.1:9010/2@58824] execute_accepting(): got exception std::system_error (error ceph::net:12, protocol aborted) at state CLOSING
INFO  2019-09-04 10:18:01,434 [shard 0] ms - [0] [osd.2(Test) v2:127.0.0.1:9010/2@58824 >> osd.4 v2:127.0.0.1:9012/4] intercepted SESSION_RECONNECT_OK_READ
WARN  2019-09-04 10:18:01,434 [shard 0] ms - [osd.2(Test) v2:127.0.0.1:9010/2@58824 >> osd.4 v2:127.0.0.1:9012/4] waiting 0.2 seconds ...
INFO  2019-09-04 10:18:01,634 [shard 0] ms - [0] [osd.2(Test) v2:127.0.0.1:9010/2 >> osd.4 v2:127.0.0.1:9012/4] intercepted SOCKET_CONNECTING
INFO  2019-09-04 10:18:01,634 [shard 0] ms - [osd.4(TestPeer) v2:127.0.0.1:9012/4 >> unknown.? -@51552] ProtocolV2::start_accept(): target_addr=127.0.0.1:51552/0
INFO  2019-09-04 10:18:01,634 [shard 0] ms - [0] [osd.2(Test) v2:127.0.0.1:9010/2 >> osd.4 v2:127.0.0.1:9012/4] intercepted BANNER_WRITE
INFO  2019-09-04 10:18:01,634 [shard 0] ms - [0] [osd.2(Test) v2:127.0.0.1:9010/2 >> osd.4 v2:127.0.0.1:9012/4] intercepted BANNER_READ
INFO  2019-09-04 10:18:01,634 [shard 0] ms - [0] [osd.2(Test) v2:127.0.0.1:9010/2 >> osd.4 v2:127.0.0.1:9012/4] intercepted BANNER_PAYLOAD_READ
INFO  2019-09-04 10:18:01,634 [shard 0] ms - [0] [osd.2(Test) v2:127.0.0.1:9010/2 >> osd.4 v2:127.0.0.1:9012/4] intercepted HELLO_WRITE
INFO  2019-09-04 10:18:01,634 [shard 0] ms - [osd.4(TestPeer) v2:127.0.0.1:9012/4 >> osd.? -@51552] UPDATE: peer_type=osd, policy(lossy=0 server=1 standby=1 resetcheck=1)
INFO  2019-09-04 10:18:01,634 [shard 0] ms - [0] [osd.2(Test) v2:127.0.0.1:9010/2 >> osd.4 v2:127.0.0.1:9012/4] intercepted HELLO_READ
INFO  2019-09-04 10:18:01,634 [shard 0] ms - [0] [osd.2(Test) v2:127.0.0.1:9010/2@51552 >> osd.4 v2:127.0.0.1:9012/4] intercepted AUTH_REQUEST_WRITE
INFO  2019-09-04 10:18:01,634 [shard 0] ms - [0] [osd.2(Test) v2:127.0.0.1:9010/2@51552 >> osd.4 v2:127.0.0.1:9012/4] intercepted AUTH_DONE_READ
INFO  2019-09-04 10:18:01,634 [shard 0] ms - [0] [osd.2(Test) v2:127.0.0.1:9010/2@51552 >> osd.4 v2:127.0.0.1:9012/4] intercepted AUTH_SIGNATURE_WRITE
INFO  2019-09-04 10:18:01,634 [shard 0] ms - [0] [osd.2(Test) v2:127.0.0.1:9010/2@51552 >> osd.4 v2:127.0.0.1:9012/4] intercepted AUTH_SIGNATURE_READ
INFO  2019-09-04 10:18:01,634 [shard 0] ms - [0] [osd.2(Test) v2:127.0.0.1:9010/2@51552 >> osd.4 v2:127.0.0.1:9012/4] intercepted SESSION_RECONNECT_WRITE
WARN  2019-09-04 10:18:01,634 [shard 0] ms - [osd.4(TestPeer) v2:127.0.0.1:9012/4 >> osd.? v2:127.0.0.1:9010/2@51552] server_reconnect: stale exsiting connect_seq exist_cs(1) < peer_cs(2), reusing existing [osd.4(TestPeer) v2:127.0.0.1:9012/4 >> osd.2 v2:127.0.0.1:9010/2@63504]
INFO  2019-09-04 10:18:01,634 [shard 0] ms - [osd.4(TestPeer) v2:127.0.0.1:9012/4 >> osd.2 v2:127.0.0.1:9010/2@63504] reconnected(replaced): gs=0, pgs=3, cs=2, client_cookie=8274171783502195587, server_cookie=1387656187526242379, in_seq=0, out_seq=0
WARN  2019-09-04 10:18:01,634 [shard 0] ms - [osd.4(TestPeer) v2:127.0.0.1:9012/4 >> osd.? v2:127.0.0.1:9010/2@51552] execute_accepting(): got exception std::system_error (error ceph::net:12, protocol aborted) at state CLOSING
INFO  2019-09-04 10:18:01,634 [shard 0] ms - [0] [osd.2(Test) v2:127.0.0.1:9010/2@51552 >> osd.4 v2:127.0.0.1:9012/4] intercepted SESSION_RECONNECT_OK_READ
INFO  2019-09-04 10:18:01,634 [shard 0] ms - [0] [osd.2(Test) v2:127.0.0.1:9010/2@51552 >> osd.4 v2:127.0.0.1:9012/4] got connected
INFO  2019-09-04 10:18:01,634 [shard 0] ms - [osd.2(Test) v2:127.0.0.1:9010/2@51552 >> osd.4 v2:127.0.0.1:9012/4] connected: gs=3, pgs=1, cs=2, client_cookie=8274171783502195587, server_cookie=1387656187526242379, in_seq=0, out_seq=0
INFO  2019-09-04 10:18:01,634 [shard 0] ms - [0] [osd.2(Test) v2:127.0.0.1:9010/2@51552 >> osd.4 v2:127.0.0.1:9012/4] ready
INFO  2019-09-04 10:18:01,634 [shard 0] ms - [0] [osd.2(Test) v2:127.0.0.1:9010/2@51552 >> osd.4 v2:127.0.0.1:9012/4] intercepted MESSAGE_READ
INFO  2019-09-04 10:18:01,634 [shard 0] ms - [0] [osd.2(Test) v2:127.0.0.1:9010/2@51552 >> osd.4 v2:127.0.0.1:9012/4] got op, pending 0 ops
INFO  2019-09-04 10:18:01,634 [shard 0] ms - Waiting for pending_send=0 pending_peer_receive=1 pending_receive=0 from [0] [osd.2(Test) v2:127.0.0.1:9010/2@51552 >> osd.4 v2:127.0.0.1:9012/4]
INFO  2019-09-04 10:18:01,634 [shard 0] ms - [0] [osd.2(Test) v2:127.0.0.1:9010/2@51552 >> osd.4 v2:127.0.0.1:9012/4] intercepted MESSAGE_WRITE
INFO  2019-09-04 10:18:01,634 [shard 0] ms - [0] [osd.2(Test) v2:127.0.0.1:9010/2@51552 >> osd.4 v2:127.0.0.1:9012/4] intercepted ACK_READ
INFO  2019-09-04 10:18:01,634 [shard 0] ms - TestPeer received op, pending 0 peer receive ops
INFO  2019-09-04 10:18:01,634 [shard 0] ms -
[SUCCESS]
INFO  2019-09-04 10:18:01,635 [shard 0] ms - [0] [osd.2(Test) v2:127.0.0.1:9010/2 >> osd.4 v2:127.0.0.1:9012/4] intercepted SOCKET_CONNECTING
WARN  2019-09-04 10:18:01,635 [shard 0] ms - [osd.2(Test) v2:127.0.0.1:9010/2 >> osd.4 v2:127.0.0.1:9012/4] waiting 0.2 seconds ...
INFO  2019-09-04 10:18:01,636 [shard 0] ms - [0] [osd.2(Test) v2:127.0.0.1:9010/2 >> osd.4 v2:127.0.0.1:9012/4] closed(closed)
WARN  2019-09-04 10:18:01,636 [shard 0] ms - [osd.2(Test) v2:127.0.0.1:9010/2 >> osd.4 v2:127.0.0.1:9012/4] wait aborted
INFO  2019-09-04 10:18:01,636 [shard 0] ms -
```

However, from the log differences I can see that the hanging case has an additional line:
`[0] [osd.2(Test) v2:127.0.0.1:9010/2@61632 >> osd.4 v2:127.0.0.1:9012/4] intercepted MESSAGE_WRITE`
which shows that after the `Interceptor` raised `abort_in_fault()`[1] in the message read operation [2], another continuation related to writes [3] was scheduled before the fault handling logic [4] of read was scheduled and changed the write state to `write_state_t::delay`. This causes the entire protocol hang somewhere. I think that's why the unit test is waiting for the connection connected/closed.

If you can still reproduce this issue, could you also enable the debug logs? Or better if you could help me find out which line of code causing the protocol hanging in the CONNECTING (probably) state?

[1] https://github.com/cyx1231st/ceph/blob/9a1ec870372721bf31552ba4bbcf6d90a931111b/src/crimson/net/ProtocolV2.cc#L140

[2] https://github.com/cyx1231st/ceph/blob/9a1ec870372721bf31552ba4bbcf6d90a931111b/src/crimson/net/ProtocolV2.cc#L1924

[3] https://github.com/cyx1231st/ceph/blob/9a1ec870372721bf31552ba4bbcf6d90a931111b/src/crimson/net/ProtocolV2.cc#L1818

[4] https://github.com/cyx1231st/ceph/blob/9a1ec870372721bf31552ba4bbcf6d90a931111b/src/crimson/net/ProtocolV2.cc#L1996

Reply:
I see, the issue is from the test, not messenger.

```
INFO  2019-09-03 20:03:26,416 [shard 0] ms - [0] [osd.2(Test) v2:127.0.0.1:9010/2@61632 >> osd.4 v2:127.0.0.1:9012/4] intercepted MESSAGE_READ
INFO  2019-09-03 20:03:26,416 [shard 0] ms - [0] [osd.2(Test) v2:127.0.0.1:9010/2@61632 >> osd.4 v2:127.0.0.1:9012/4] intercepted MESSAGE_WRITE
INFO  2019-09-03 20:03:26,447 [shard 0] ms - TestPeer received op, pending 0 peer receive ops
INFO  2019-09-03 20:03:26,447 [shard 0] ms - Waiting for connection [0] [osd.2(Test) v2:127.0.0.1:9010/2 >> osd.4 v2:127.0.0.1:9012/4] connected/closed
```
Assumed steps from the logs (`Test` <--> `TestPeer`):

Line 1: `Test` tries to read a message, was intercepted and fault raised;
Line 2. `Test` writes a message;
Line 3. The message was received by `TestPeer`, indicating the message write from `Test` is successful;
Line 3-4. The fault raised from Line.1 was handled, and because `Test` has nothing to send, its connection state switches to STANDBY;
Line 4. The test suite checks the state of the connection. `Test` will be at STANDBY state forever because no more pending messages. So the test suite will hang there. 

The fix should be simple, just remove the cases with `Breakpoint{Tag::MESSAGE, false}`.

Reply:
fixed.